Fixed #32996 -- Cached PathInfos on relations.

kezabelle opened the pull request:
[Ticket is 32996](https://code.djangoproject.com/ticket/32996) (currently un-triaged and thus not accepted, but let's at least make sure the tests are in an OK shape. Guarantee that one of the linters will chastise me, at the very least).

kezabelle commented on django/db/models/fields/reverse_related.py:
```diff
@@ -138,6 +138,16 @@ def __eq__(self, other):
+        # PathInfo may contain an intermediate M2M table that's been
+        # dynamically created and doesn't exist in the .models module.
+        state.pop('path_infos', None)
+        return state
```
This is implemented because otherwise deep down in a query's `alias_map` a `Join` can include a rel one side of which is an internal m2m intermediate table (`<class 'queryset_pickle.models.M2MModel_groups'>`). Tested by `tests.queryset_pickle.tests.PickleabilityTestCase.test_filter_reverse_fk` and `...test_pickle_filteredrelation`

kezabelle commented on django/db/models/fields/related.py:
```diff
@@ -495,6 +495,12 @@ def __init__(self, to, on_delete, from_fields, to_fields, rel=None, related_name
+        for cached_prop in ('path_infos', 'reverse_path_infos'):
+            state.pop(cached_prop, None)
+        return state
+
```
This isn't strictly necessary; no tests failed without it being here, but it would seem prudent to perhaps do something like this for all the touched areas, for consistency and to guard against there being some way in which an unpickleable value ends up in there.

kezabelle commented:
Yeah, I'll give it another once over/rebase/tidy pass and then it can be looked at for problems.

felixxm commented:
> Yeah, I'll give it another once over/rebase/tidy pass and then it can be looked at for problems.

Thanks :+1:

kezabelle commented:
OK I've updated everything and tidied up where necessary, and fixed some issues along the way.

`__getstate__` is implemented on `ManyToOneRel` already, and without removing the cached attributes causes errors in `queryset_pickle.tests.PickleabilityTestCase.test_filter_reverse_fk` and `queryset_pickle.tests.PickleabilityTestCase.test_pickle_filteredrelation`. Because any of the `ForeignObjectRel` subclasses _could_ include the same, I've implemented `ForeignObjectRel.__getstate__` to allow them all to work for downstream users.

Where I previously also had `__getstate__` implemented on `ForeignObject`, when adding tests, I noticed it's never used (heh). Because it inherits from `Field`, which has a custom implementation of `__reduce__`, pickling and unpickling an `ForeignObject` will always keep the cached attributes due to reloading from a Model's internal Meta.

But! `ForeignObjectRel` doesn't have a custom `__reduce__`, so `__getstate__` is _also_ used when doing `copy.copy(fo_rel)` - so a shallow copy of a rel object deletes the cached values. Thus, for _some_ consistency, I've implemented `ForeignObject.__copy__` to do the same. 

So:
- pickling a `ForeignObjectRel` deletes the attributes. (`__getstate__`)
- copying a `ForeignObjectRel` deletes the attributes. (`__getstate__`)
- copying a `ForeignObject` deletes the attributes. (`__copy__`)
- _deep_ copying a `ForeignObject` deletes the attributes, + those of the backing `ForeignObjectRel` (`Field.__deepcopy__` + methods above).
- pickling a `ForeignObject` does _not_ delete the attributes, because they're still held in memory by the `Model` (`Field.__reduce__`).

kezabelle commented on django/db/models/fields/reverse_related.py:
```diff
@@ -138,6 +138,16 @@ def __eq__(self, other):
+        state = self.__dict__.copy()
+        # We delete the cached_property here because it can be re-calculated at
+        # first invocation after loads() is used.
+        # We need to remove it for ubclasses like ManyToOneRel where the
```
Good catch, thanks.

felixxm reviewed (commented):
@kezabelle Thanks :+1: I left two small comments. State/copy changes looks good, but I need to really focus :brain: and review them again in detail :mag:

kezabelle commented on django/db/models/fields/related.py:
```diff
@@ -1541,11 +1556,11 @@ def _get_path_info(self, direct=False, filtered_relation=None):
-            join1infos = linkfield2.get_reverse_path_info()
-            join2infos = linkfield1.get_path_info(filtered_relation)
+            join1infos = linkfield2.reverse_path_infos
+            join2infos = linkfield1.get_path_info(filtered_relation) if filtered_relation else linkfield1.path_infos
```
Based on the way it's _currently_ implemented, I can't see how. You'd end up at maximum recursion depth I think, off the top of my head (`get_path_info(None)` -> `path_infos` -> `get_path_info(None)` -> ...)

We could do away with the cached property entirely, and manage the caching of the result directly as an internal attribute by doing `self._path_infos = self._get_path_info(...)`. That may put more onus on downstream consumers of a `ManyToManyField` with a customised implementation of `get_path_info` or `_get_path_info` which would potentially have to juggle the cache themselves (though this is perhaps entirely hypothetical), the benefit (if there is one) of using a separate attribute would be that any inheritor of something which has `get_path_info` would automatically get the caching, though that naturally comes at the cost of their `get_path_info` implementation being stable on repeated calls (which you'd think it is, but who knows!)

kezabelle commented on django/db/models/sql/query.py:
```diff
@@ -1528,8 +1528,8 @@ def names_to_path(self, names, opts, allow_many=True, fail_on_missing=False):
-            if hasattr(field, 'get_path_info'):
-                pathinfos = field.get_path_info(filtered_relation)
+            if hasattr(field, 'path_infos'):
+                pathinfos = field.get_path_info(filtered_relation) if filtered_relation else field.path_infos
```
You know, I briefly remembered that I'd probably need to do that, as you mentioned the preference in another comment elsewhere. Then I entirely forgot about it. Will update it :)

felixxm reviewed (commented):
@kezabelle Thanks again :+1:

I pushed small edits. Also, I found that `ManyToManyField.reverse_path_infos` was not covered by tests so I changed `ForeignObjectRel.get_path_info()` to the:
```python
     def get_path_info(self, filtered_relation=None):
        if filtered_relation:
            return self.field.get_reverse_path_info(filtered_relation)
        else:
            return self.field.reverse_path_infos
```

adamchainz commented:
A worthy optimization, thank you @kezabelle and @felixxm .
